File: src/lib/delivery-register-state.ts
export type DeliveryRegisterStateItem = { label: string; required: boolean; status: string };

export type DeliveryRegisterStateEntry = {
  manifest: {
    items: DeliveryRegisterStateItem[];
    readiness: {
      clientNetworkAccepted: boolean;
      facilityDispatched: boolean;
      deadlineRisk: "on_track" | "at_risk" | "overdue";
      hasDeliveryContactGaps: boolean;
    };
  } | null;
};

/** The concise operational status used by the Delivery register. */
export function getDeliveryRegisterState(entry: DeliveryRegisterStateEntry) {
  if (!entry.manifest) return "not_configured";
  const { readiness } = entry.manifest;
  if (readiness.clientNetworkAccepted) return "accepted";
  if (readiness.deadlineRisk !== "on_track" || readiness.hasDeliveryContactGaps || entry.manifest.items.some((item) => ["qc_failed", "rejected"].includes(item.status))) return "needs_attention";
  if (readiness.facilityDispatched) return "dispatched";
  return "in_progress";
}

/** One clear next step prevents a technical delivery register from becoming a wall of statuses. */
export function getNextDeliveryAction(entry: DeliveryRegisterStateEntry) {
  if (!entry.manifest) return "Apply a delivery profile";
  const { readiness } = entry.manifest;
  if (readiness.hasDeliveryContactGaps) return "Choose a delivery recipient";
  const blocked = entry.manifest.items.find((item) => item.required && ["qc_failed", "rejected"].includes(item.status));
  if (blocked) return `Resolve ${blocked.label}`;
  const waiting = entry.manifest.items.find((item) => item.required && item.status === "dispatched");
  if (waiting) return `Confirm receipt for ${waiting.label}`;
  const outstanding = entry.manifest.items.find((item) => item.required && !["receipt_confirmed", "waived"].includes(item.status));
  return outstanding ? `Prepare ${outstanding.label}` : "Delivery complete";
}

File: backend/app/delivery_register_state.py
"""Concise delivery-register status and next-action helpers."""

from __future__ import annotations

from typing import TypedDict


class DeliveryItem(TypedDict):
    label: str
    required: bool
    status: str


class DeliveryReadiness(TypedDict):
    client_network_accepted: bool
    facility_dispatched: bool
    deadline_risk: str
    has_delivery_contact_gaps: bool


def delivery_register_state(manifest: dict[str, object] | None) -> str:
    if not manifest:
        return "not_configured"
    readiness = manifest["readiness"]
    assert isinstance(readiness, dict)
    items = manifest["items"]
    assert isinstance(items, list)
    if readiness["client_network_accepted"]:
        return "accepted"
    if (
        readiness["deadline_risk"] != "on_track"
        or readiness["has_delivery_contact_gaps"]
        or any(item["status"] in {"qc_failed", "rejected"} for item in items)
    ):
        return "needs_attention"
    return "dispatched" if readiness["facility_dispatched"] else "in_progress"


def next_delivery_action(manifest: dict[str, object] | None) -> str:
    if not manifest:
        return "Apply a delivery profile"
    readiness = manifest["readiness"]
    items = manifest["items"]
    assert isinstance(readiness, dict) and isinstance(items, list)
    if readiness["has_delivery_contact_gaps"]:
        return "Choose a delivery recipient"
    blocked = next((item for item in items if item["required"] and item["status"] in {"qc_failed", "rejected"}), None)
    if blocked:
        return f"Resolve {blocked['label']}"
    waiting = next((item for item in items if item["required"] and item["status"] == "dispatched"), None)
    if waiting:
        return f"Confirm receipt for {waiting['label']}"
    outstanding = next(
        (item for item in items if item["required"] and item["status"] not in {"receipt_confirmed", "waived"}), None
    )
    return f"Prepare {outstanding['label']}" if outstanding else "Delivery complete"
